s3transfer/manager.py

```diff
@@ -0,0 +1,165 @@
+
+
+MB = 1024 * 1024
+
+
+class TransferConfig(object):
+    def __init__(self,
+                 multipart_threshold=8 * MB,
+                 max_concurrency=10,
+                 multipart_chunksize=8 * MB,
+                 max_queue_size=0):
+        """Configurations for the transfer mangager
+
+        :param multipart_threshold: The threshold for which multipart
+            transfers occur.
+
+        :param max_concurrency: The maximum number or requests that
+            can happen at a time.
+
+        :param multipart_chunksize: The size of each transfer if a request
+            becomes a multipart transfer.
+
+        :param max_queue_size: The maximum amount of requests that
+            can be queued at a time. A value of zero means that there
+            is no maximum.
+        """
+        self.multipart_threshold = multipart_threshold
+        self.max_concurrency = max_concurrency
+        self.multipart_chunksize = multipart_chunksize
+        self.max_queue_size = max_queue_size
+
+
+class TransferManager(object):
+    VALID_SUBSCRIBER_TYPES = [
+        'queued',
+        'progress',
+        'done'
+    ]
+
+    ALLOWED_UPLOAD_ARGS = [
+        'ACL',
+        'CacheControl',
+        'ContentDisposition',
+        'ContentEncoding',
+        'ContentLanguage',
+        'ContentType',
+        'Expires',
+        'GrantFullControl',
+        'GrantRead',
+        'GrantReadACP',
+        'GrantWriteACL',
+        'Metadata',
+        'RequestPayer',
+        'ServerSideEncryption',
+        'StorageClass',
+        'SSECustomerAlgorithm',
+        'SSECustomerKey',
+        'SSECustomerKeyMD5',
+        'SSEKMSKeyId',
+    ]
+
+    def __init__(self, client, config=None):
+        """A transfer manager interface for Amazon S3
+
+        :param client: Client to be used by the manager
+        :param config: TransferConfig to associate specific configurations
+        """
+        self._client = client
+        self._config = config
+        if config is None:
+            self._config = TransferConfig()
+        self._osutil = OSUtils()
+        self._executor = BoundedExecutor(
+            max_size=self._config.max_queue_size,
+            max_num_threads=self._config.max_concurrency
+        )
+
+    def upload(self, fileobj, bucket, key, extra_args=None, subscribers=None):
+        """Uploads a file to S3
+
+        :type fileobj: str
+        :param fileobj: The name of a file to upload.
+
+        :type bucket: str
+        :param bucket: The name of the bucket to upload to
+
+        :type key: str
+        :param key: The name of the key to upload to
+
+        :type extra_args: dict
+        :param extra_args: Extra arguments that may be passed to the
+            client operation
+
+        :type subscribers: a list of subscribers
+        :param subscribers: The list of subscribers to be invoked in the
+            order provided based on the event emit during the process of
+            the transfer request.
```

Comment:
I think it would be worth having a base class that a user can implement that would serve as documentation for the valid notifications you can subscribe to.